server/src/main/java/org/elasticsearch/reservedstate/ReservedClusterStateHandler.java

```diff
@@ -60,6 +62,25 @@
      */
     TransformState transform(Object source, TransformState prevState) throws Exception;
 
+    /**
+     * A transformation method for handlers that need to perform transformations outside of the cluster state
+     *
+     * <p>
+     * All changes to the cluster state should be done now, but if the handlers need to perform other state modifications,
+     * like role mappings or ingest pipelines, they should implement the modification logic in this method.
+     *
+     * We loop through the postTransform method calls after all transform methods are called.
+     *
+     * @param source The parsed information specific to this handler from the combined cluster state content
+     * @param clusterState The final cluster state from all handler calls to transform
+     * @param previousKeys The previous keys saved for this handler
+     * @return The current keys set by this handler
+     * @throws Exception
+     */
+    default Set<String> postTransform(Object source, ClusterState clusterState, Set<String> previousKeys) throws Exception {
```

Comment:
I think the signature of this method be simplifed to
```
Set<String> postTransform(Object source, TransformState prevState)
```
since ClusterState + Keys = TransformState?

The contract of this method seems a bit odd to me. I almost feel it should be returned as a runnable/callable for calling `transform`. The main reason is what needs to be done with `postTransform` (update, delete) is already known at `transform` time. The content is also known. `postTransform` only needs to perform the actions that are outside of clusterstate. So we don't really need to pass it the clusterstate and keys again. In fact, passing them again here raises the question about whether they should be same or different compared to those passed into `transform`. The name `previousKeys` is also a bit fuzzy because it can be current or previous?

That said, the current code works for now and I haven't involved in the design of file based state feature. So you don't have to do anything. Mainly food for thoughts. Thanks!

Reply:
I agree, I'll simplify the method interface. I like your idea of callables, let me explore it a bit to see if it simplifies the whole thing.

Reply:
I ended up using your idea for the post transform, looks much nicer. I added a post transform supplier option in the transform state, which is then simply called when the cluster state is saved.